okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java

```diff
@@ -38,10 +40,45 @@
 public final class HttpHeaders {
-  private static final String TOKEN = "([^ \"=]*)";
-  private static final String QUOTED_STRING = "\"([^\"]*)\"";
-  private static final Pattern PARAMETER
-      = Pattern.compile(" +" + TOKEN + "=(:?" + QUOTED_STRING + "|" + TOKEN + ") *(:?,|$)");
+  // regexes according to RFC 7235
+  private static final String TOKEN_PATTERN_PART = "[!#$%&'*+.^_`|~\\p{Alnum}-]+";
+  private static final String TOKEN68_PATTERN_PART = "[\\p{Alnum}._~+/-]+=*";
+  private static final String OWS_PATTERN_PART = "[ \\t]*";
+  private static final String QUOTED_PAIR_PATTERN_PART = "\\\\([\\t \\p{Graph}\\x80-\\xFF])";
+  private static final String QUOTED_STRING_PATTERN_PART =
+          "\"(?:[\\t \\x21\\x23-\\x5B\\x5D-\\x7E\\x80-\\xFF]|" + QUOTED_PAIR_PATTERN_PART + ")*\"";
+  private static final String AUTH_PARAM_PATTERN_PART = TOKEN_PATTERN_PART + OWS_PATTERN_PART + '='
+          + OWS_PATTERN_PART + "(?:" + TOKEN_PATTERN_PART + '|' + QUOTED_STRING_PATTERN_PART + ')';
+  private static final String CHALLENGE_PATTERN_PART = TOKEN_PATTERN_PART + "(?: +(?:"
+          + TOKEN68_PATTERN_PART + "|(?:,|" + AUTH_PARAM_PATTERN_PART + ")(?:" + OWS_PATTERN_PART
+          + ",(?:" + OWS_PATTERN_PART + AUTH_PARAM_PATTERN_PART + ")?)*)?)?";
+  private static final String PROXY_AUTHENTICATE_VALUE_SPLIT_PATTERN_PART =
+          "(?:" + OWS_PATTERN_PART + ',' + OWS_PATTERN_PART + ")+";
+
+  private static final Pattern PROXY_AUTHENTICATE_VALUE_PATTERN = Pattern.compile("^(?:,"
+          + OWS_PATTERN_PART + ")*" + CHALLENGE_PATTERN_PART + "(?:" + OWS_PATTERN_PART + ",(?:"
+          + OWS_PATTERN_PART + CHALLENGE_PATTERN_PART + ")?)*$");
+  private static final Pattern AUTH_SCHEME_PATTERN =
+          Pattern.compile('^' + TOKEN_PATTERN_PART + '$');
+  private static final Pattern AUTH_SCHEME_AND_TOKEN68_PATTERN =
+          Pattern.compile('^' + TOKEN_PATTERN_PART + " +" + TOKEN68_PATTERN_PART + '$');
+  private static final Pattern AUTH_SCHEME_AND_PARAM_PATTERN =
+          Pattern.compile('^' + TOKEN_PATTERN_PART + " +" + AUTH_PARAM_PATTERN_PART + '$');
+  private static final Pattern AUTH_PARAM_PATTERN =
+          Pattern.compile('^' + AUTH_PARAM_PATTERN_PART + '$');
+  private static final Pattern TOKEN_PATTERN = Pattern.compile('^' + TOKEN_PATTERN_PART + '$');
+  private static final Pattern QUOTED_PAIR_PATTERN = Pattern.compile(QUOTED_PAIR_PATTERN_PART);
+
+  private static final Pattern PROXY_AUTHENTICATE_VALUE_SPLIT_PATTERN =
+          Pattern.compile(PROXY_AUTHENTICATE_VALUE_SPLIT_PATTERN_PART);
+  private static final Pattern WHITESPACE_SPLIT_PATTERN = Pattern.compile(" +");
+  private static final Pattern AUTH_PARAM_SPLIT_PATTERN =
+          Pattern.compile(OWS_PATTERN_PART + '=' + OWS_PATTERN_PART);
+  private static final Pattern QUOTED_STRING_AUTH_PARAM_AT_END_PATTERN =
+          Pattern.compile(TOKEN_PATTERN_PART + OWS_PATTERN_PART + '=' + OWS_PATTERN_PART
+                  + QUOTED_STRING_PATTERN_PART + '$');
```

Comment:
The time to init this class just went up ... is there a way to do this with less code?

Reply:
Well, the class init is only done once per JVM run, so I thought this is not too much of an issue, don't you?